[View in Colaboratory](https://colab.research.google.com/github/todnewman/coe_training/blob/master/Basic_Data_Engineering.ipynb)

# Basic Pandas Data Engineering
Author: W. Tod Newman

## Learning Objectives


*   Learn how to import files into Pandas Dataframes and how to develop understanding of what's in the data
*   Learn how to access and manipulate the data in the DataFrame
*   Learn how to do vectorized math and Boolean Filtering on the Dataframe.

In this exercise we're using data from the UC Irvine Open Data archive.  This data is in .CSV format and is easily pulled into Pandas.  See the Handling Complex Data files notebook for details on how to manipulate larger, zipped files from UCI.

In [0]:
import pandas as pd
import numpy as np

# Import a Data File and Read it Into a Pandas DataFrame

Here  we pull data from the UC Irvine open data repository.  This data is useful because often it can be used to test algorithms and ML workflows.  Plus, it works well in CoLaboratory because we can download it through the firewall.

First we read the file in as a .CSV and then we print the dataframe.  You can print the dataframe by merely typing "df" only if you do it on the last line.  Otherwise, you need to print(df).  The formatting of the latter will be less pleasant than the former, BTW.

You can comment out any of the last 4 lines to see what those lines do.  The last one uncommented will be printed.

In [0]:
df = pd.read_csv(r'https://archive.ics.uci.edu/ml/machine-learning-databases/parkinsons/parkinsons.data')
df
df[0:10]
df.head()
#df.tail(2)

name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0  phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1  phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2  phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3  phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4  phon_R01_S01_5      116.014       141.781       110.655         0.01284   

   MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer    ...     \
0           0.00007   0.00370   0.00554     0.01109       0.04374    ...      
1           0.00008   0.00465   0.00696     0.01394       0.06134    ...      
2           0.00009   0.00544   0.00781     0.01633       0.05233    ...      
3           0.00009   0.00502   0.00698     0.01505       0.05492    ...      
4           0.00011   0.00655   0.00908     0.01966       0.06425    ...      

   Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0      0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1      0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2      0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3      0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4      0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   

    spread2        D2       PPE  
0  0.266482  2.301442  0.284654  
1  0.335590  2.486855  0.368674  
2  0.311173  2.342259  0.332634  
3  0.334147  2.405554  0.368975  
4  0.234513  2.332180  0.410335  

[5 rows x 24 columns]

## Print Column Headers (keys)

This is a useful thing to do.  If you assign df.keys() to a variable, it becomes a Python list.  You can iterate through the list to access the individual keys one by one or you can access each parameter in the list with it's index (i.e., keys[3]).  This is a useful way to select features without having to know the name of the header.

In [0]:
keys = df.keys()
print (keys)

Index([u'name', u'MDVP:Fo(Hz)', u'MDVP:Fhi(Hz)', u'MDVP:Flo(Hz)',
       u'MDVP:Jitter(%)', u'MDVP:Jitter(Abs)', u'MDVP:RAP', u'MDVP:PPQ',
       u'Jitter:DDP', u'MDVP:Shimmer', u'MDVP:Shimmer(dB)', u'Shimmer:APQ3',
       u'Shimmer:APQ5', u'MDVP:APQ', u'Shimmer:DDA', u'NHR', u'HNR', u'status',
       u'RPDE', u'DFA', u'spread1', u'spread2', u'D2', u'PPE'],
      dtype='object')


## Information on the DataFrame

Pandas gives us a couple of useful calls to get info on the different features in the DataFrame.

In [0]:
print (df.info())
print (df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 24 columns):
name                195 non-null object
MDVP:Fo(Hz)         195 non-null float64
MDVP:Fhi(Hz)        195 non-null float64
MDVP:Flo(Hz)        195 non-null float64
MDVP:Jitter(%)      195 non-null float64
MDVP:Jitter(Abs)    195 non-null float64
MDVP:RAP            195 non-null float64
MDVP:PPQ            195 non-null float64
Jitter:DDP          195 non-null float64
MDVP:Shimmer        195 non-null float64
MDVP:Shimmer(dB)    195 non-null float64
Shimmer:APQ3        195 non-null float64
Shimmer:APQ5        195 non-null float64
MDVP:APQ            195 non-null float64
Shimmer:DDA         195 non-null float64
NHR                 195 non-null float64
HNR                 195 non-null float64
status              195 non-null int64
RPDE                195 non-null float64
DFA                 195 non-null float64
spread1             195 non-null float64
spread2             195 non-null floa

## Accessing parameters in the DataFrame

You can see from the below how Pandas gives access to the index, columns, and values of the DataFrame.  You can use .columns to change the header names if that's desirable.  Understanding the shape of the dataframe (in this case, it is 195 rows by 24 columns) will help you understand if you get shape mismatch errors.

In [0]:
print (df.index)
print ("Shape of the Dataframe is: ",df.shape)
#print (df.columns)
#print (df.values)

RangeIndex(start=0, stop=195, step=1)
('Shape of the Dataframe is: ', (195, 24))


## Sorting

sort_values is useful.  You can either sort by something like I did here (keys[3]) or you can sort by the header name ('MDVP:Flo(Hz)').  

In [0]:
print (keys[1:6])
df.sort_values(keys[3], ascending=False)

Index([u'MDVP:Fo(Hz)', u'MDVP:Fhi(Hz)', u'MDVP:Flo(Hz)', u'MDVP:Jitter(%)',
       u'MDVP:Jitter(Abs)'],
      dtype='object')


name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
170  phon_R01_S42_6      244.990       272.210       239.170         0.00451   
167  phon_R01_S42_3      260.105       264.919       237.303         0.00339   
43   phon_R01_S10_2      241.404       248.834       232.483         0.00281   
44   phon_R01_S10_3      243.439       250.912       232.435         0.00210   
46   phon_R01_S10_5      245.510       262.090       231.848         0.00235   
166  phon_R01_S42_2      237.323       243.709       229.256         0.00303   
45   phon_R01_S10_4      242.852       255.034       227.911         0.00225   
42   phon_R01_S10_1      237.226       247.326       225.227         0.00298   
63   phon_R01_S17_4      228.832       234.619       223.634         0.00296   
64   phon_R01_S17_5      229.401       252.221       221.156         0.00205   
169  phon_R01_S42_5      240.301       245.135       219.783         0.00517   
62   phon_R01_S17_3      222.236       231.345       205.495         0.00266   
111  phon_R01_S26_3      208.519       220.315       199.020         0.00609   
33   phon_R01_S07_4      202.266       211.604       197.079         0.00180   
34   phon_R01_S07_5      203.184       211.526       196.160         0.00178   
35   phon_R01_S07_6      201.464       210.565       195.708         0.00198   
32   phon_R01_S07_3      198.383       215.203       193.104         0.00212   
31   phon_R01_S07_2      199.228       209.512       192.091         0.00241   
30   phon_R01_S07_1      197.076       206.896       192.055         0.00289   
112  phon_R01_S26_4      204.664       221.300       189.621         0.00841   
113  phon_R01_S26_5      210.141       232.706       185.258         0.00534   
47   phon_R01_S10_6      252.455       261.487       182.786         0.00185   
40   phon_R01_S08_5      186.163       197.724       177.584         0.00298   
149  phon_R01_S35_4      202.632       565.740       177.258         0.01627   
38   phon_R01_S08_3      180.198       201.249       175.456         0.00284   
152  phon_R01_S35_7      198.116       233.099       174.478         0.02714   
39   phon_R01_S08_4      187.733       202.324       173.015         0.00316   
151  phon_R01_S35_6      192.818       224.429       168.793         0.03107   
36   phon_R01_S08_1      177.876       192.921       168.013         0.00411   
41   phon_R01_S08_6      184.055       196.537       166.977         0.00258   
..              ...          ...           ...           ...             ...   
90   phon_R01_S21_7      166.605       206.008        78.032         0.00742   
194  phon_R01_S50_6      214.289       260.277        77.973         0.00567   
25   phon_R01_S06_2      104.400       206.002        77.968         0.00633   
29   phon_R01_S06_6      162.568       198.346        77.630         0.00502   
160  phon_R01_S39_2      114.238       124.393        77.022         0.00581   
24   phon_R01_S06_1      163.656       200.841        76.779         0.00742   
93   phon_R01_S22_3      152.125       161.469        76.596         0.00382   
13   phon_R01_S04_2      139.173       179.139        76.556         0.00390   
14   phon_R01_S04_3      152.845       163.305        75.836         0.00294   
101  phon_R01_S24_5      128.451       150.449        75.632         0.01551   
17   phon_R01_S04_6      168.778       232.181        75.603         0.00718   
26   phon_R01_S06_3      171.041       208.313        75.501         0.00455   
103  phon_R01_S25_1      150.258       154.609        75.349         0.00248   
84   phon_R01_S21_1      176.858       205.560        75.344         0.00747   
0    phon_R01_S01_1      119.992       157.302        74.997         0.00784   
193  phon_R01_S50_5      198.764       396.961        74.904         0.00740   
88   phon_R01_S21_5      173.898       211.350        74.677         0.00448   
192  phon_R01_S50_4      174.688       240.005        74.287         0.01360   
116  phon_R01_S

## Print out selected Columns in the DataFrame

This allows you to pick which columns you want to see.  The first line selects just one column.  The second shows how you can select a slice (multiple contiguous columnts). The third shows how you can just select discrete sets of columns.  The fourth line shows us passing an array of headers to the DataFrame.  You will probably notice that this isn't much different than the approach for the 3rd line.

In [0]:
df[keys[1]]
df[keys[1:6]]
df[['MDVP:Fo(Hz)', 'MDVP:Jitter(%)']]
my_header_list = ['MDVP:Fo(Hz)', 'MDVP:Jitter(%)', 'MDVP:PPQ']
df[my_header_list]

MDVP:Fo(Hz)  MDVP:Jitter(%)  MDVP:PPQ
0        119.992         0.00784   0.00554
1        122.400         0.00968   0.00696
2        116.682         0.01050   0.00781
3        116.676         0.00997   0.00698
4        116.014         0.01284   0.00908
5        120.552         0.00968   0.00750
6        120.267         0.00333   0.00202
7        107.332         0.00290   0.00182
8         95.730         0.00551   0.00332
9         95.056         0.00532   0.00332
10        88.333         0.00505   0.00330
11        91.904         0.00540   0.00336
12       136.926         0.00293   0.00153
13       139.173         0.00390   0.00208
14       152.845         0.00294   0.00149
15       142.167         0.00369   0.00203
16       144.188         0.00544   0.00292
17       168.778         0.00718   0.00387
18       153.046         0.00742   0.00432
19       156.405         0.00768   0.00399
20       153.848         0.00840   0.00450
21       153.880         0.00480   0.00267
22       167.930         0.00442   0.00247
23       173.917         0.00476   0.00258
24       163.656         0.00742   0.00390
25       104.400         0.00633   0.00375
26       171.041         0.00455   0.00234
27       146.845         0.00496   0.00275
28       155.358         0.00310   0.00176
29       162.568         0.00502   0.00253
..           ...             ...       ...
165      236.200         0.00277   0.00153
166      237.323         0.00303   0.00159
167      260.105         0.00339   0.00186
168      197.569         0.00803   0.00448
169      240.301         0.00517   0.00283
170      244.990         0.00451   0.00237
171      112.547         0.00355   0.00190
172      110.739         0.00356   0.00200
173      113.715         0.00349   0.00203
174      117.004         0.00353   0.00218
175      115.380         0.00332   0.00199
176      116.388         0.00346   0.00213
177      151.737         0.00314   0.00162
178      148.790         0.00309   0.00186
179      148.143         0.00392   0.00231
180      150.440         0.00396   0.00233
181      148.462         0.00397   0.00235
182      149.818         0.00336   0.00198
183      117.226         0.00417   0.00270
184      116.848         0.00531   0.00346
185      116.286         0.00314   0.00192
186      116.556         0.00496   0.00263
187      116.342         0.00267   0.00148
188      114.563         0.00327   0.00184
189      201.774         0.00694   0.00396
190      174.188         0.00459   0.00259
191      209.516         0.00564   0.00292
192      174.688         0.01360   0.00564
193      198.764         0.00740   0.00390
194      214.289         0.00567   0.00317

[195 rows x 3 columns]

In [0]:
df[1:4] # Filter out 3 rows in the DataFrame
df.loc[:,[keys[1], keys[7]]] # Filter headers 1 and 7 for all records
df.loc[1:4,[keys[1], keys[7]]] # Filter headers 1 and y for just three records

MDVP:Fo(Hz)  MDVP:PPQ
1      122.400   0.00696
2      116.682   0.00781
3      116.676   0.00698
4      116.014   0.00908

## Boolean Indexing

Often its useful to be able to select all records that meet a criteria.  For example, we might want all records where MDVP:Fo(Hz) > 120.  Or as another example, we might all records where YEAR == 2018.

In [0]:
mdvp_hi_fo_freq = df[keys[1]] > 120  # Set up the boolean.  The vaue of mdvp_hi_fo_freq will be True if it is > 120 and False if it is < 120
df[mdvp_hi_fo_freq]

name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
1    phon_R01_S01_2      122.400       148.650       113.819         0.00968   
5    phon_R01_S01_6      120.552       131.162       113.787         0.00968   
6    phon_R01_S02_1      120.267       137.244       114.820         0.00333   
12   phon_R01_S04_1      136.926       159.866       131.276         0.00293   
13   phon_R01_S04_2      139.173       179.139        76.556         0.00390   
14   phon_R01_S04_3      152.845       163.305        75.836         0.00294   
15   phon_R01_S04_4      142.167       217.455        83.159         0.00369   
16   phon_R01_S04_5      144.188       349.259        82.764         0.00544   
17   phon_R01_S04_6      168.778       232.181        75.603         0.00718   
18   phon_R01_S05_1      153.046       175.829        68.623         0.00742   
19   phon_R01_S05_2      156.405       189.398       142.822         0.00768   
20   phon_R01_S05_3      153.848       165.738        65.782         0.00840   
21   phon_R01_S05_4      153.880       172.860        78.128         0.00480   
22   phon_R01_S05_5      167.930       193.221        79.068         0.00442   
23   phon_R01_S05_6      173.917       192.735        86.180         0.00476   
24   phon_R01_S06_1      163.656       200.841        76.779         0.00742   
26   phon_R01_S06_3      171.041       208.313        75.501         0.00455   
27   phon_R01_S06_4      146.845       208.701        81.737         0.00496   
28   phon_R01_S06_5      155.358       227.383        80.055         0.00310   
29   phon_R01_S06_6      162.568       198.346        77.630         0.00502   
30   phon_R01_S07_1      197.076       206.896       192.055         0.00289   
31   phon_R01_S07_2      199.228       209.512       192.091         0.00241   
32   phon_R01_S07_3      198.383       215.203       193.104         0.00212   
33   phon_R01_S07_4      202.266       211.604       197.079         0.00180   
34   phon_R01_S07_5      203.184       211.526       196.160         0.00178   
35   phon_R01_S07_6      201.464       210.565       195.708         0.00198   
36   phon_R01_S08_1      177.876       192.921       168.013         0.00411   
37   phon_R01_S08_2      176.170       185.604       163.564         0.00369   
38   phon_R01_S08_3      180.198       201.249       175.456         0.00284   
39   phon_R01_S08_4      187.733       202.324       173.015         0.00316   
..              ...          ...           ...           ...             ...   
145  phon_R01_S34_6      223.361       263.872        87.638         0.00352   
146  phon_R01_S35_1      169.774       191.759       151.451         0.01568   
147  phon_R01_S35_2      183.520       216.814       161.340         0.01466   
148  phon_R01_S35_3      188.620       216.302       165.982         0.01719   
149  phon_R01_S35_4      202.632       565.740       177.258         0.01627   
150  phon_R01_S35_5      186.695       211.961       149.442         0.01872   
151  phon_R01_S35_6      192.818       224.429       168.793         0.03107   
152  phon_R01_S35_7      198.116       233.099       174.478         0.02714   
153  phon_R01_S37_1      121.345       139.644        98.250         0.00684   
156  phon_R01_S37_4      122.336       142.369        94.794         0.00727   
158  phon_R01_S37_6      126.144       154.284        97.543         0.00975   
159  phon_R01_S39_1      127.930       138.752       112.173         0.00605   
165  phon_R01_S42_1      236.200       244.663       102.137         0.00277   
166  phon_R01_S42_2      237.323       243.709       229.256         0.00303   
167  phon_R01_S42_3      260.105       264.919       237.303         0.00339   
168  phon_R01_S42_4      197.569       217.627        90.794         0.00803   
169  phon_R01_S42_5      240.301       245.135       219.783         0.00517   
170  phon_R01_S42_6      244.990       272.210       239.170         0.00451   
177  phon_R01_S

## Math

In [0]:
print("Mean Value for Column 1 is %s" % df[keys[1]].mean())  # Calculate the mean value for the whole column
print("Median Value for Column 1 is %s" % df[keys[1]].median())
#print("Cumulative Sum for column 1 is %s" % df[keys[1]].apply(np.cumsum))
df['Product1'] = df[keys[1]] * df[keys[2]]  #Multiply two DataFrame columns into a product.
df['Product1']

Mean Value for Column 1 is 154.22864102564102
Median Value for Column 1 is 148.79


0      18874.981584
1      18194.760000
2      15298.293702
3      16086.236796
4      16448.580934
5      15811.841424
6      16505.924148
7      12218.674880
8      12642.869640
9      11416.510768
10      9914.495920
11     10649.008384
12     21889.811916
13     24931.312047
14     24960.352725
15     30914.924985
16     50358.956692
17     39187.044818
18     26909.925134
19     29622.794190
20     25498.459824
21     26599.696800
22     32447.602530
23     33519.892995
24     32868.834696
25     21506.608800
26     35630.063833
27     30646.698345
28     35325.768114
29     32244.712528
           ...     
165    57789.400600
166    57837.751007
167    68906.756495
168    42996.348763
169    58906.185635
170    66688.727900
171    15010.843578
172    12579.618183
173    13241.315745
174    16903.099864
175    14204.316420
176    15018.474744
177    28860.984348
178    23562.235610
179    23107.641426
180    24588.064040
181    23913.962036
182    24482.808106
183    14527.232050
